src/CommandRunner.cpp

```diff
@@ -196,124 +202,85 @@ class Timer {
-  pid_t pid = popen2(command, args);
-
-  auto waitProcess = [pid, &status, loggingMetadata, command,
-                      &forceKillRequired, &exitCode, &signalCode]() {
-    int rc = waitpid(pid, &status, WNOHANG);
-    if (rc < 0) {
-      TASK_LOG(ERROR, loggingMetadata)
-          << "Error when waiting for child process running the "
-          << "external command: " << strerror(errno);
-      forceKillRequired = true;
-      return true;
-    }
-
-    if (rc > 0 && (WIFEXITED(status) || WIFSIGNALED(status))) {
-      if (WIFEXITED(status) && WEXITSTATUS(status) != 0) {
-        exitCode = WEXITSTATUS(status);
-        TASK_LOG(ERROR, loggingMetadata)
-            << "Failed to successfully run the command \"" << command
-            << "\", it failed with status " << std::to_string(exitCode) << ".";
-      }
-      if (WIFSIGNALED(status) && WTERMSIG(status) != 0) {
-        signalCode = WTERMSIG(status);
-        TASK_LOG(ERROR, loggingMetadata)
-            << "Failed to successfully run the command \"" << command
-            << "\", it exited with signal " << std::to_string(signalCode)
-            << ".";
-      }
-      return true;
-    }
-    return false;
-  };
-
-  milliseconds tickPeriod(100);
-  Timer t1(tickPeriod, milliseconds(timeoutInSeconds * 1000));
-  t1.run(waitProcess);
-
-  if (t1.hasTimedOut() && !forceKillRequired) {
-    TASK_LOG(WARNING, loggingMetadata)
-        << "External command took too long to exit. "
-        << "Sending SIGTERM...";
-    if (kill(pid, SIGTERM) == -1) {
-      TASK_LOG(ERROR, loggingMetadata) << "Failed to send SIGTERM: "
-                                       << strerror(errno);
-      forceKillRequired = true;
-    } else {
-      Timer t(tickPeriod, milliseconds(1000));
-      t.run(waitProcess);
-      if (t.hasTimedOut()) {
-        forceKillRequired = true;
-      }
-    }
-  }
-
-  if (forceKillRequired) {
-    TASK_LOG(WARNING, loggingMetadata)
-        << "External command is still running. Sending SIGKILL...";
-    if (kill(pid, SIGKILL) == -1) {
-      TASK_LOG(ERROR, loggingMetadata) << "Failed to kill the command: "
-                                       << strerror(errno);
-      return Error("Command \"" + command +
-                   "\" took too long to execute and SIGKILL failed.");
-    } else {
-      return Error("Command \"" + command + "\" took too long to execute.");
-    }
-  }
-
-  if (exitCode > 0) {
-    return Error("Command \"" + command + "\" exited with return code " +
-                 std::to_string(exitCode) + ".");
-  } else if (signalCode > 0) {
-    return Error("Command \"" + command + "\" exited via signal " +
-                 std::to_string(signalCode) + ".");
+Future<Try<bool>> runCommandWithTimeout(
+    const std::string& theCommand, const std::vector<std::string>& args,
+    unsigned long timeoutInSeconds, const logging::Metadata& loggingMetadata) {
+  // FIXME: remove this
+  vector<string> full_args;
+  full_args.push_back(theCommand);
```

Comment:
isn't there a method to push everything at once ?

Reply:
Sadly, no. It's C++, the standard lib is not really designed to make your code more digest...